openfoodfacts/images.py

```diff
@@ -254,3 +254,99 @@ def download_image(
+        image ID to the detail about the image:
+        - `uploaded_t`: the upload timestamp
+        - `uploader`: the username of the uploader
+        - `sizes`: dictionary mapping image size (`100`, `200`, `400`, `full`)
+            to the information about each resized image:
+            - `h`: the height of the image
+            - `w`: the width of the image
+            - `url`: the URL of the image
+    - `selected` contains the images that are selected, and maps the
+        image key (`nutrition`, `ingredients`, `packaging`, or `front`) to
+        a dictionary mapping the language to the selected image details.
+        The selected image details are the following fields:
+        - `imgid`: the image ID
+        - `rev`: the revision ID
+        - `sizes`: dictionary mapping image size (`100`, `200`, `400`, `full`)
+            to the information about each resized image:
+            - `h`: the height of the image
+            - `w`: the width of the image
+            - `url`: the URL of the image
+        - `generation`: information about how to generate the selected image
+            from the uploaded image:
+            - `geometry`
+            - `x1`, `y1`, `x2`, `y2`: the coordinates of the crop
+            - `angle`: the rotation angle of the selected image
+            - `coordinates_image_size`: 400 or "full", indicates if the
+                geometry coordinates are relative to the full image, or to a
+                resized version (max width and max height=400)
+            - `normalize`: indicates if colors should be normalized
+            - `white_magic`: indicates if the background is white and should
+                be removed (e.g. photo on a white sheet of paper)
+
+    See https://github.com/openfoodfacts/openfoodfacts-server/pull/11818
+    for more details.
+    """
+
+    if not is_new_image_schema(images):
+        return images
+
+    images_with_legacy_schema = {}
+
+    for image_id, image_data in images["uploaded"].items():
+        images_with_legacy_schema[image_id] = {
+            "sizes": {
+                # remove URL field
+                size: {k: v for k, v in image_size_data.items() if k != "url"}
+                for size, image_size_data in image_data["sizes"].items()
+            },
+            "uploaded_t": image_data["uploaded_t"],
+            "uploader": image_data["uploader"],
+        }
+
+    for selected_key, image_by_lang in images["selected"].items():
+        for lang, image_data in image_by_lang.items():
+            new_image_data = {
+                "imgid": image_data["imgid"],
+                "rev": image_data["rev"],
+                "sizes": {
+                    # remove URL field
+                    size: {k: v for k, v in image_size_data.items() if k != "url"}
+                    for size, image_size_data in image_data["sizes"].items()
+                },
+                **image_data["generation"],
```

Comment:
note that generation fields can be absent. e.g. if there's no cropping, then there won't be x1, y1, x2, y2. Not sure if that could break some stuff.